File: dotnet-desktop-guide/winforms/advanced/snippets/how-to-enable-binaryformatter-clipboard-support/vb/ClipboardUsage.vb
Imports System
Imports System.Reflection.Metadata
Imports System.Windows.Forms

Namespace ClipboardExamples
    Public Class ClipboardUsageExample

        Public Class MyCustomType
            Public Property Data As String
        End Class

        ' <ClipboardUsage>
        ' Use the resolver with clipboard operations
        Private Shared Function SecureTypeResolver(typeName As TypeName) As Type
            ' Implementation from SecureTypeResolver example
            ' ... (allow-list implementation here)
            Throw New InvalidOperationException($"Type '{typeName.FullName}' is not permitted")
        End Function

        Public Shared Sub UseSecureTypeResolver()
            ' Retrieve legacy data using the secure type resolver
            Dim data As MyCustomType = Nothing
            If Clipboard.TryGetData("LegacyData", AddressOf SecureTypeResolver, data) Then
                ProcessLegacyData(data)
            Else
                Console.WriteLine("No compatible data found on clipboard")
            End If
        End Sub
        ' </ClipboardUsage>

        Private Shared Sub ProcessLegacyData(data As MyCustomType)
            Console.WriteLine($"Processing legacy data: {data.Data}")
        End Sub

    End Class
End Namespace

File: dotnet-desktop-guide/winforms/advanced/snippets/how-to-enable-binaryformatter-clipboard-support/csharp/ClipboardUsage.cs
using System;
using System.Reflection.Metadata;
using System.Windows.Forms;

namespace ClipboardExamples
{
    public class ClipboardUsageExample
    {
        public class MyCustomType
        {
            public string Data { get; set; }
        }

        // <ClipboardUsage>
        // Use the resolver with clipboard operations
        private static Type SecureTypeResolver(TypeName typeName)
        {
            // Implementation from SecureTypeResolver example
            // ... (allow-list implementation here)
            throw new InvalidOperationException($"Type '{typeName.FullName}' is not permitted");
        }

        public static void UseSecureTypeResolver()
        {
            // Retrieve legacy data using the secure type resolver
            if (Clipboard.TryGetData("LegacyData", SecureTypeResolver, out MyCustomType data))
            {
                ProcessLegacyData(data);
            }
            else
            {
                Console.WriteLine("No compatible data found on clipboard");
            }
        }
        // </ClipboardUsage>

        private static void ProcessLegacyData(MyCustomType data)
        {
            Console.WriteLine($"Processing legacy data: {data.Data}");
        }
    }
}